### PR TITLE
fix(parser) Fix freezing issue with illegal 0 width matches

### LINKED ISSUE #2522: Infinite loop for illegal tokens when `ignore_illegals` or `highlightAuto`
On 10.0.1, the following XML:

```xml
<?xml version="1.0"?>
<response value="ok" xml:lang="en">
  <text>Ok</text>
  <comment html_allowed="true"/>
  <ns1:description><![CDATA[
  CDATA is <not> magical.
  ]]></ns1:description>
  <a></a> <a/>
</response>
```

When passed through `highlightAuto` or `highlight` with `ignore_illegals` turned on, will cause an infinite loop, specifically, because of `'$'` as an illegal in CSS selector attributes:

https://github.com/highlightjs/highlight.js/blob/e8624ccb65cfd859a0273b58ec3897afd16a5498/src/languages/css.js#L72

### PR DESCRIPTION
**Pull request by yyyc514**
Closes #2522

### CODE REVIEW

**Comment by yyyc514** [comment]
@allejo @egor-rogov Will publish this tomorrow if no one reports an issue until then.

**Comment by wooorm** [comment]
Thanks Josh for the quick fix! ✨

**Comment by yyyc514** [comment]
10.0.2 being released now.

File: src/highlight.js [hunk 1]
```diff
@@ -365,6 +365,23 @@ const HLJS = function(hljs) {
...
+      }
+
+      // infinite loops are BAD, this is a last ditch catch all. if we have a
+      // decent number of iterations yet our index (cursor position in our
+      // parsing) still 3x behind our index then something is very wrong
+      // so we bail
+      if (iterations > 100000 && iterations > match.index * 3) {
+        const err = new Error('potential infinite loop, way more iterations than matches');
+        throw err;
+      }
```
Comment (egor-rogov) [inline src/highlight.js:383]: Sorry I'm late to the party. Quick fix was the right thing to do.
I'd move this check out to the `for (;;)` loop, in order not to deal with `iterations` inside `processLexeme` function.

Comment (yyyc514) [reply src/highlight.js:383]: Except then we'd literally have to check it ALL the time... where-as right now it's more of a fallback... but I'll play around with it.

Comment (yyyc514) [reply src/highlight.js:383]: But technically your right in that if there were a bug in the above code that this wouldn't be well suited to catch it... :-/

**Pull request merged**